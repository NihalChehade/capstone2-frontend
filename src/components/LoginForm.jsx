import React, { useState, useContext  } from "react";
import { Form, FormGroup, Label, Input, Button, Alert, Spinner } from "reactstrap";
import UserContext from "../UserContext";
import { useNavigate, Navigate } from "react-router-dom";
import "./LoginForm.css"

const LoginForm = () => {
  const { currentUser, login, errorMessages} = useContext(UserContext);
  const [formData, setFormData] = useState({
    username: "",
    password: ""
  });

  const navigate = useNavigate();
  const [formError, setFormError] = useState("");
  const [loading, setLoading] = useState(false);

  // Handle input changes
  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value
    }));
    // Clear the specific field error when user changes the input
    if (formError[name]) {
      setFormError({ ...formError, [name]: undefined });
    }
  };

  // Handle form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    setLoading(true); // Start spinner
    try{
      const success = await login(formData);
      if (success) {
        navigate("/"); // Redirect after successful login
      }
    } catch(error){
      setFormError(error);
    } finally{
      setLoading(false);
    }
     
  };

  if(currentUser){
    return <Navigate to="/"/>;
  }

  return (
    <div className="Login">
      <h1>Log In</h1>
      
      {loading && (
        <div className="text-center">
          <Spinner color="primary" />
        </div>
      )} {/* Display spinner when loading is true */}

      {/* Display error message if there's an error */}
       {formError.general && <Alert color="danger" >{formError.general}</Alert>}

      <Form  onSubmit={handleSubmit}>
        <FormGroup>
          <Label for="username">Username</Label>
          <Input
            type="text"
            name="username"
            id="username"
            value={formData.username}
            onChange={handleChange}
            required
          />
        </FormGroup>

        <FormGroup>
          <Label for="password">Password</Label>
          <Input
            type="password"
            name="password"
            id="password"
            value={formData.password}
            onChange={handleChange}
            required
          />
        </FormGroup>

        <Button type="submit" color="primary" block disabled={loading}>
          Submit
        </Button>
      </Form>
    </div>
  );
};

export default LoginForm;
