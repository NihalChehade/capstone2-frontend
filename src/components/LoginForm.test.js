import React from 'react';
import { render, fireEvent, waitFor, act } from '@testing-library/react';
import { BrowserRouter, useNavigate  } from 'react-router-dom';
import UserContext from "../UserContext";
import LoginForm from './LoginForm';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

describe('LoginForm', () => {
  let mockNavigate;
  const mockLogin = jest.fn();
  
  beforeEach(() => {
    mockNavigate = jest.fn();
    useNavigate.mockReturnValue(mockNavigate);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  const providerProps = {
    currentUser: null,
    login: mockLogin
  };

  const Wrapper = ({ children }) => (
    <UserContext.Provider value={providerProps}>
      <BrowserRouter>{children}</BrowserRouter>
    </UserContext.Provider>
  );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders without crashing and matches snapshot', () => {
    const { asFragment } = render(<LoginForm />, { wrapper: Wrapper });
    expect(asFragment()).toMatchSnapshot();
  });

  it('submits the form with entered values', async () => {
    mockLogin.mockResolvedValue(true);
    const { getByLabelText, getByRole } = render(<LoginForm />, { wrapper: Wrapper });

    fireEvent.change(getByLabelText(/username/i), { target: { value: 'testuser' } });
    fireEvent.change(getByLabelText(/password/i), { target: { value: 'password123' } });
    fireEvent.click(getByRole('button', { name: /submit/i }));

    await waitFor(() => expect(mockLogin).toHaveBeenCalledWith({
      username: 'testuser',
      password: 'password123'
    }));
    expect(mockNavigate).toHaveBeenCalledWith("/");
  });

  it('displays error message on failed login', async () => {
   
    mockLogin.mockRejectedValueOnce({general:'Invalid credentials'});
    const { getByLabelText, getByRole, findByText } = render(<LoginForm />, { wrapper: Wrapper });

    fireEvent.change(getByLabelText(/username/i), { target: { value: 'testuser' } });
    fireEvent.change(getByLabelText(/password/i), { target: { value: 'password' } });
    fireEvent.click(getByRole('button', { name: /submit/i }));

    const errorMessage = await findByText(/invalid credentials/i);
    expect(errorMessage).toBeInTheDocument();
  });

  it('redirects if user is already logged in', () => {
    const { container } = render(
      <UserContext.Provider value={{ ...providerProps, currentUser: { name: 'John Doe' } }}>
        <BrowserRouter>
          <LoginForm />
        </BrowserRouter>
      </UserContext.Provider>
    );
    expect(container.innerHTML).toBe("");
  });
});
