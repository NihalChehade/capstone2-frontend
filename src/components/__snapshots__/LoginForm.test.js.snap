// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders without crashing and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Login"
  >
    <h1>
      Log In
    </h1>
     
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="username"
        >
          Username
        </label>
        <input
          class="form-control"
          id="username"
          name="username"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="password"
        >
          Password
        </label>
        <input
          class="form-control"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary d-block w-100"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
